refextract/references/pdf.py

```diff
@@ -0,0 +1,103 @@
+# License, or (at your option) any later version.
+#
+# refextract is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with refextract; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+import re
+
+from PyPDF2 import PdfFileReader
+
+from .regexs import re_reference_in_dest
+
+
+def extract_texkeys_from_pdf(pdf_file):
+    """
+    Extract the texkeys from the given PDF file
+
+    This is done by looking up the named destinations in the PDF
+
+    @param pdf_file: path to a PDF
+
+    @return: list of all texkeys found in the PDF
+    """
+    with open(pdf_file, 'rb') as pdf_stream:
+        pdf = PdfFileReader(pdf_stream, strict=False)
+        destinations = pdf.getNamedDestinations()
+        # not all named destinations point to references
+        refs = [dest for dest in destinations.iteritems()
+                if re_reference_in_dest.match(dest[0])]
+        if _destinations_in_two_columns(pdf, refs):
+            print("* PDF: Using two-column layout")
+
+            def sortfunc(dest_couple):
+                return _destination_position(pdf, dest_couple[1])
+
+        else:
+            print("* PDF: Using single-column layout")
+
+            def sortfunc(dest_couple):
+                (page, _, ypos, xpos) = _destination_position(pdf,
+                                                              dest_couple[1])
+                return (page, ypos, xpos)
+
+        refs.sort(key=sortfunc)
+        # extract the TeXkey from the named destination name
+        return [re_reference_in_dest.match(destname).group(1)
+                for (destname, _) in refs]
+
+
+def _destinations_in_two_columns(pdf, destinations, cutoff=3):
+    """
+    Check if the named destinations are organized along two columns (heuristic)
+
+    @param pdf: a PdfFileReader object
+    @param destinations:
+
+    'cutoff' is used to tune the heuristic: if 'cutoff' destinations in the
+    would-be second column start at the same position, return True
+    """
+    # iterator for the x coordinates of refs in the would-be second column
+    xpositions = (_destination_position(pdf, dest)[3] for (_, dest)
+                  in destinations
+                  if _destination_position(pdf, dest)[1] == 1)
+    xpos_count = {}
+    for xpos in xpositions:
+        xpos_count[xpos] = xpos_count.get(xpos, 0) + 1
+        if xpos_count[xpos] >= cutoff:
+            return True
+    return False
```

Comment:
> I know, but this would break compatibility with Python 2.6 that does not have collections.Counter, and the plan is to backport this to legacy.

Aww, ok.

> Plus `Counter` would build the whole dict while here we return early.

TBH this matters only if it is code in tight loop which is expected to run as fast as possible. In all other cases, I would call this a Premature Optimization: http://wiki.c2.com/?PrematureOptimization